compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -34,7 +34,6 @@ use super::{
     Pointer, Projectable, Scalar, ValueVisitor,
 };
 
-// for the validation errors
```

Comment:
Why does this comment have to go? Can't we have separate import blocks separated by comments any more?

It's not a super important comment, but being able to apply custom grouping via comments seems quite valuable so it would be concerning for rustfmt to just flat-out ignore that.  Previously on Zulip it was said that rustmft would only regroup within a contiguous sequence of `use` statements, which led me to believe it wouldn't regroup across comments here?

Reply:
> rustmft would only regroup within a contiguous sequence of use statements

Depends what you mean by "contiguous sequence". Blank lines and comments don't act as separators between use statements. Other items (e.g. `mod`, type declarations) do.

Reply:
> Blank lines and comments don't act as separators between use statements.

For blank lines that makes sense since the entire point is for rustfmt to do the grouping. But for comments I find this highly disruptive.

> And "for the validation errors" seems like a low-value comment to me.

As the maintainer and one of the main authors of this file, I disagree. The imports that carry this comment are unusual, usually we'd not have `*` imports of an enum at a global level, nor `as` aliases. So it's worth explaining why we do something strange like that.

Reply:
I find it rather concerning that we should entirely lose the ability to add a comment to a group of `use` statements. The standard library also has several of these cases show up in this PR, where comments that are helpful to structure the file are now being erased because we have a tool that can't deal with them.

Reply:
I could take or leave this particular comment, but I do agree that commented groups are useful and would be a shame to lose. Is there any rustfmt setting we can use (or request) to keep that ability?

Reply:
No. rustfmt will, with `group_imports = "StdExternalCrate"`, put all the `use` declarations into three sections, with a single blank line between them. The whole concept of a single comment applying to multiple lines is suspect within this model. That's the whole point of the first commit in this PR, to get rid of such comments.

The example in question looks like this:
```
use super::{
    err_ub, format_interp_error, machine::AllocMap, throw_ub, AllocId, AllocKind, CheckInAllocMsg,
    GlobalAlloc, ImmTy, Immediate, InterpCx, InterpResult, MPlaceTy, Machine, MemPlaceMeta, OpTy,
    Pointer, Projectable, Scalar, ValueVisitor,
};

// for the validation errors
use super::InterpError::UndefinedBehavior as Ub; 
use super::InterpError::Unsupported as Unsup;
use super::UndefinedBehaviorInfo::*;
use super::UnsupportedOpInfo::*;
```
After reformatting, it looks like this:
```
use super::machine::AllocMap;
// for the validation errors
use super::InterpError::UndefinedBehavior as Ub;
use super::InterpError::Unsupported as Unsup;
use super::UndefinedBehaviorInfo::*;
use super::UnsupportedOpInfo::*;
use super::{
    err_ub, format_interp_error, throw_ub, AllocId, AllocKind, CheckInAllocMsg, GlobalAlloc, ImmTy,
    Immediate, InterpCx, InterpResult, MPlaceTy, Machine, MemPlaceMeta, OpTy, Pointer, Projectable,
    Scalar, ValueVisitor,
};
```
The comment sticks to the `use super::InterpError::UndefinedBehavior as Ub;` line, which is a reasonable choice and would be fine if the comment only applied to that line. But rustfmt can't know that it actually applies to four lines. (That fact isn't even perfectly clear to a human reader.) Maybe you could argue that the blank line before the comment should make it clear. But the formatting will be getting rid of unnecessary blank lines. Maybe you could argue that a blank line followed by a comment should be interpreted as the comment applying to multiple lines. Perhaps? But at this point it's worth mentioning that the number of comments on use items in the codebase is actually really small, and the number of comments that apply to multiple use items is even smaller. This PR deals with all of them. It's a couple of dozen comments out of many thousands of use items.

In summary, rustfmt does just fine with like 99.99% of cases and for 0.01% of cases we have a slightly suboptimal experience. It's a standard autoformatting trade-off: we accept the small number of suboptimal cases for all the other benefits, such as consistency and not having to make decisions about how to style things. For this particular case, I could do something like this:
```
use super::machine::AllocMap;
use super::InterpError::UndefinedBehavior as Ub; // for the validation errors
use super::InterpError::Unsupported as Unsup; // for the validation errors
use super::UndefinedBehaviorInfo::*; // for the validation errors
use super::UnsupportedOpInfo::*; // for the validation errors
use super::{
    err_ub, format_interp_error, throw_ub, AllocId, AllocKind, CheckInAllocMsg, GlobalAlloc, ImmTy,
    Immediate, InterpCx, InterpResult, MPlaceTy, Machine, MemPlaceMeta, OpTy, Pointer, Projectable,
    Scalar, ValueVisitor,
};
```
which is totally unambiguous, more so than the original comment.

Reply:
I think having a group of code with empty lines before and after it, and a comment at the top of it, makes it fairly clear to a human reader that the comment applies to the entire group. But rustfmt is frequently bad at dealing with comments and likes to move them around in ways that change their meaning (https://github.com/rust-lang/rustfmt/issues/3994, https://github.com/rust-lang/rustfmt/issues/5485), so it is no surprise that this happens here, too.  It is a constant frustration of mine that issues like that are ignored in the name of the Grand Goal Of Consistency. For basic rustfmt formatting the advantages it provides are so big that the overall tradeoff is beneficial (though it is unfortunate that fixing these rustfmt issues is nobody's priority), but for `use` statements I am not convinced that there's a significant enough problem here worth solving.

> In summary, rustfmt does just fine with like 99.99% of cases and for 0.01% of cases we have a slightly suboptimal experience

That "one in 10000" number is complete guesswork on your end. A couple dozen cases in around 3k source files also shows that you are off by around 2 orders of magnitude.

Reply:
There are something like 18,000 use items in `compiler/` and `library/` alone. Maybe the correct number is 99.9% or 99.5%.  No matter what the exact number is, we're having an argument about suboptimal handling of one or two actual comments in the entire codebase. Is that worth blocking this entire change from happening? In my opinion, no.